### ISSUE TITLE
[web-animations-2] [scroll-animations-1] Auto-aligning the start time of an animation associated with a scroll timeline

### ISSUE THREAD

**Issue by graouts** [w3c/csswg-drafts#10965]
I'm working on implementing [Scroll-driven Animations](https://drafts.csswg.org/scroll-animations-1/) in WebKit and having some trouble understanding how the start time of an animation associated with a scroll timeline ever gets resolved.

From my understanding, the start time of such an animation is set by the [auto-aligning the start time procedure](https://drafts.csswg.org/web-animations-2/#auto-aligning-start-time). Per [HTML Processing Model: Event loop](https://drafts.csswg.org/scroll-animations-1/#event-loop):

> When updating timeline current time, the [start time](https://drafts.csswg.org/web-animations-1/#animation-start-time) of any attached animation is conditionally updated. For each attached animation, run the procedure for [calculating an auto-aligned start time](https://drafts.csswg.org/web-animations-2/#calculating-an-auto-aligned-start-time).

So each time animations are updating, we have a chance to auto-align the start time of scroll-driven animations. But, the [auto-aligning the start time procedure] step 3 says:

> If [play state](https://drafts.csswg.org/web-animations-1/#animation-play-state) is idle, abort this procedure.

The [play state is idle](https://drafts.csswg.org/web-animations-1/#play-states) when:

> All of the following conditions are true:
> 
> - The [current time](https://drafts.csswg.org/web-animations-1/#animation-current-time) of animation is [unresolved](https://drafts.csswg.org/web-animations-1/#unresolved), and
> - the [start time](https://drafts.csswg.org/web-animations-1/#animation-start-time) of animation is [unresolved](https://drafts.csswg.org/web-animations-1/#unresolved), and
> - animation does not have either a [pending play task](https://drafts.csswg.org/web-animations-1/#pending-play-task) or a [pending pause task](https://drafts.csswg.org/web-animations-1/#pending-pause-task),

My assumption is that the start time is unresolved since we're yet to set it via the _auto-align the start time_ procedure. We do not have any pending tasks since when initially playing that scroll-driven animation, we aborted the [play procedure](https://drafts.csswg.org/web-animations-2/#playing-an-animation-section) after step 10:

> If the following three conditions are all satisfied:
> 
> - animation’s [hold time](https://drafts.csswg.org/web-animations-1/#animation-hold-time) is [unresolved](https://drafts.csswg.org/web-animations-1/#unresolved), and
> - aborted pause is false, and
> - animation does not have a [pending playback rate](https://drafts.csswg.org/web-animations-1/#pending-playback-rate),
> 
> abort this procedure.

And the current time is unresolved precisely we don't yet have a start time and return unresolved in this step of the [calculating the current time of an animation procedure](https://drafts.csswg.org/web-animations-1/#the-current-time-of-an-animation):

> If any of the following are true:
> 
> - the animation has no associated [timeline](https://drafts.csswg.org/web-animations-1/#timeline), or
> - the associated [timeline](https://drafts.csswg.org/web-animations-1/#timeline) is [inactive](https://drafts.csswg.org/web-animations-1/#inactive-timeline), or
> - the animation’s [start time](https://drafts.csswg.org/web-animations-1/#animation-start-time) is [unresolved](https://drafts.csswg.org/web-animations-1/#unresolved),
>
> The [current time](https://drafts.csswg.org/web-animations-1/#animation-current-time) is an [unresolved](https://drafts.csswg.org/web-animations-1/#unresolved) time value.

Therefore it is unclear to me when an animation associated with a scroll timeline can ever have a resolved start time or current time. What am I missing?

**Comment by johannesodland** [comment]
I think this is the same issue I ran into in  https://github.com/w3c/csswg-drafts/issues/9871. I'm closing that issue as a duplicate in favor of yours.

**Comment by kevers-google** [comment]
I believe this section is missing a condition:

https://www.w3.org/TR/web-animations-1/#waiting-for-the-associated-effect

Presently the conditions for being ready are that: 
* User agent has completed any setup...
* Timeline is active.

Adding a third condition that the animation has a start or hold time, resolves things.  This keeps the animation in a play pending state until we have a resolved (auto-aligned) start time.

Previously, the timeline being active was sufficient, but now we potentially need a layout update to resolve the view timeline range and calculate an auto-aligned start time. This effectively leaves the animation not in effect while waiting on the animation to be ready, but this prevents a potential flicker by initializing with one start time only to jump to another after the view timeline range has been updated.

**Comment by graouts** [comment]
Note that Web Animations Level 2 already has an [update for this section](https://drafts.csswg.org/web-animations-2/#waiting-for-the-associated-effect) that matches what you are saying:

> the animation’s [hold time](https://drafts.csswg.org/web-animations-1/#animation-hold-time) or [start time](https://drafts.csswg.org/web-animations-1/#animation-start-time) is resolved.

**Comment by kevers-google** [comment]
Ah yes.  I see a TODO in the Blink implementation to add a condition to prevent aborting if playing a scroll driven animation.  Apologies for not following up on that one.

In Blink, we have an extra condition to avoid an early abort if using a scroll timeline.  This will require clarification in the spec.  Hold time is expected to be set if using a document timeline, but not if using a scroll timeline.  Looking at the play procedure in web-animations-2, I, I see an issue in the "if seek time is resolved" step.  This has us setting the start time if using a finite-timeline, which worked prior to the introduction of view-timelines. As we don't necessarily know at the time of play what the start time should be for a scroll-timeline, we need to defer the calculation, and avoid an early abort if expecting a deferred start time.  I'll follow up with Rob on this.

**Comment by graouts** [comment]
> Ah yes. I see a TODO in the Blink implementation to add a condition to prevent aborting if playing a scroll driven animation. Apologies for not following up on that one.

I implement the "auto-aligning the start time" procedure in WebKit in [285176@main](https://commits.webkit.org/285176@main) and made a change to step 10 of the [playing an animation](https://drafts.csswg.org/web-animations-2/#playing-an-animation-section) procedure such as if the _auto align start time_ flag is true and _start time_ is unresolved we proceed with the rest of the procedure instead of bailing out.

See https://github.com/WebKit/WebKit/commit/433599a15892af435851c20280e58a0cc91596cb#diff-ac012fc9792c68f4fdf706f4021da22aa0be5d72e7547cb42e3a9c3f36df6d18R1182.
